File: 프로그래머스/2/92335. k진수에서 소수 개수 구하기/k진수에서 소수 개수 구하기.cpp
#include <string>
#include <vector>
#include <cmath>
#include <sstream>

using namespace std;

// 소수 판별 함수
bool isPrime(long long num) {
    if (num < 2) return false;
    for (long long i = 2; i <= sqrt(num); ++i) {
        if (num % i == 0) return false;
    }
    return true;
}

string toKBase(int n, int k) {
    string result = "";
    while (n > 0) {
        result = to_string(n % k) + result;
        n /= k;
    }
    return result;
}

int solution(int n, int k) {
    string k_base = toKBase(n, k); // 진법 변환
    stringstream ss(k_base);
    string segment;
    int count = 0;

    while (getline(ss, segment, '0')) { // '0'을 기준으로 분리
        if (!segment.empty()) {
            long long num = stoll(segment);
            if (isPrime(num)) count++;
        }
    }
    return count;
}

File: 프로그래머스/2/92335. k진수에서 소수 개수 구하기/k진수에서 소수 개수 구하기.cs
using System;
using System.Collections.Generic;

public class Solution 
{
    int answer;
    public int solution(int n, int k) 
    {
        answer = 0;
        int num = n;
        string result = Solve01(num, k);
        Solve02(result);
        return answer;
    }
    
    string Solve01(int num, int k)
    {
        List<int> list = new List<int>();
        while(num > 0)
        {
            list.Add(num % k);
            num /= k;
        }
        
        string result = "";
        for(int i=list.Count - 1; i>=0; i--)
        {
            result += list[i].ToString();
        }
        return result;
    }
    
    void Solve02(string str)
    {
        string tmp = "";
        foreach(char c in str)
        {
            if(c != '0')
            {
                tmp += c;
            }
            else
            {
                if(tmp.Length > 0 && tmp != "1") 
                {
                    if(IsPrime(tmp))
                    {
                        answer++;
                    }
                }
                tmp = "";
            }
        }
        
        if(tmp.Length > 0 && tmp != "1") 
        {
            if(IsPrime(tmp))
            {
                answer++;
            }
        }
    }
    
    bool IsPrime(string s)
    {
        long num = long.Parse(s);
        for(long i=2; i * i <= num; i++)
        {
            if(num % i == 0)
            {
                return false;
            }
        }
        return true;
    }
}